packages/reason-relay/reason-relay-ppx/library/Query.re

```diff
@@ -0,0 +1,356 @@
+               ],
+              ~onResult:
+                 Belt.Result.t(
+                   [%t typeFromGeneratedModule(["Types", "response"])],
+                   Js.Exn.t,
+                 ) =>
+                 unit,
+              ~networkCacheConfig: option(ReasonRelay.cacheConfig)=?,
+              ~fetchPolicy: option(ReasonRelay.fetchQueryFetchPolicy)=?,
+              (),
+            )
+            : unit => {
+          let _ =
+            Internal.internal_fetchQuery(
+              environment,
+              [%e valFromGeneratedModule(["node"])],
+              variables->[%e
+                           valFromGeneratedModule([
+                             "Internal",
+                             "convertVariables",
+                           ])
+                         ],
+              Some({
+                networkCacheConfig,
+                fetchPolicy: fetchPolicy->ReasonRelay.mapFetchQueryFetchPolicy,
+              }),
+            )
+            ->ReasonRelay.Observable.(
+                subscribe(
+                  makeObserver(
+                    ~next=
+                      res =>
+                        onResult(
+                          Ok(
+                            res->[%e
+                                   valFromGeneratedModule([
+                                     "Internal",
+                                     "convertResponse",
+                                   ])
+                                 ],
+                          ),
+                        ),
+                    ~error=err => onResult(Error(err)),
+                    (),
+                  ),
+                )
+              );
+          ();
+        }
+      ],
+      [%stri
+        let fetchPromised =
+            (
+              ~environment: ReasonRelay.Environment.t,
+              ~variables: [%t
+                 typeFromGeneratedModule(["Types", "variables"])
+               ],
+              ~networkCacheConfig: option(ReasonRelay.cacheConfig)=?,
+              ~fetchPolicy: option(ReasonRelay.fetchQueryFetchPolicy)=?,
+              (),
+            )
+            : Promise.t(
+                Belt.Result.t(
+                  [%t typeFromGeneratedModule(["Types", "response"])],
+                  Js.Exn.t,
+                ),
+              ) => {
+          let (promise, resolve) = Promise.pending();
+
+          let _ =
+            Internal.internal_fetchQuery(
+              environment,
+              [%e valFromGeneratedModule(["node"])],
+              variables->[%e
+                           valFromGeneratedModule([
+                             "Internal",
+                             "convertVariables",
+                           ])
+                         ],
+              Some({
+                networkCacheConfig,
+                fetchPolicy: fetchPolicy->ReasonRelay.mapFetchQueryFetchPolicy,
+              }),
+            )
+            ->ReasonRelay.Observable.(
```

Comment:
relay observable has a function `toPromise` which we could use instead of subscribing.
https://github.com/facebook/relay/blob/9ef158d08bc6d1f98c352aa6e1b610807e2c779f/packages/relay-runtime/network/RelayObservable.js#L428

Reply:
This is actually deliberately not using that because I want to never reject and instead return a result type. Now whether that's a _good idea_ is another discussion 😉

Reply:
Perhaps users will expect the underlying code will call `toPromise`. Not sure, I don't have a strong opinion about this.